### ISSUE TITLE
[2.1.x] Service resolve raise a warning

### ISSUE THREAD

**Issue by ogarbe** [phalcon/cphalcon#11709]
Hello,

These lines raises a warning in 2.1.x : https://github.com/phalcon/cphalcon/blob/2.1.x/phalcon/di/service.zep#L173-L178

I initialize all my code with error Exception when warning or notice : 

``` php
        /**
         * Set generic error handler
         */
        set_error_handler(
            function ($errno, $errstr, $errfile, $errline) {
                if (0 === error_reporting()) {
                    return false;
                }
                throw new \ErrorException($errstr, 0, $errno, $errfile, $errline);
            },
            E_ALL
        );

        /**
         * Set generic exception handler
         */
        set_exception_handler('handleException');
```

and when trying to resolve a closure, here is the error i face : 

``` shell
Fatal error: Uncaught exception 'ErrorException' with message 'Cannot bind an instance to a static closure'

ErrorException: Cannot bind an instance to a static closure in /home/olivier.garbé/www/lib_phalcon/Library/Loader.php on line 134

Call Stack:
    0.0467     714272   1. handleException() 
    0.0467     714416   2. Phalcon\Di->get()
    0.0467     714512   3. Phalcon\Di\Service->resolve() 
    0.0467     714648   4. Closure::bind()
```

In 2.0.x these lines didn't exists, hence no error.

**Comment by andresgutierrez** [comment]
This happens because of this:

> Phalcon\Di is now bound to services closures allowing use Phalcon\Di as $this to access services within them

https://github.com/phalcon/cphalcon/blob/2.1.x/CHANGELOG.md

https://github.com/phalcon/cphalcon/issues/11029#issuecomment-200612702

**Comment by ogarbe** [comment]
thanks andres, but how can we get rid of the warning ?

**Comment by andresgutierrez** [comment]
Some service is returning a closure which was created in a static context thus generating the exception. To avoid this you have to avoid return a static-created closure to the di->resolve method.

**Comment by ogarbe** [comment]
@andresgutierrez 

https://docs.phalconphp.com/en/latest/reference/di.html 

the use os static closure is in the documentation for Phalcon, it

```php
// Using an anonymous function
$di->set('logger', function () {
    return new LoggerFile('../apps/logs/error.log');
});
```

how would you change this (which is very handy) ?

**Comment by andresgutierrez** [comment]
@ogarbe That example works for me in 2.1.x

**Comment by sergeyklay** [comment]
@ogarbe ping

**Comment by ogarbe** [comment]
my pb is that i have a factory default which is called by a static method called initialize, which use closure.

Hence i've done this commit https://github.com/vpg/cphalcon/commit/e217b35ba6a989f758b4fa29837f834c8fcafe75 which fix the error and add backwards compatibility. 

The fix just don't remove a warning, which is better, and doesn't change anything else.

**Comment by piit79** [comment]
@sergeyklay @ogarbe Is the fix for this going into 3.0.2? I can confirm @ogarbe 's fix works correctly for us. Most of the time we use static closures for DI service definitions so we were getting **lots** of warnings in PHPUnit and were forced to set `convertWarningsToExceptions = false`.

**Comment by Jurigag** [comment]
This is not merged because reflection api is heavy. There was supposed to be created some zephir thing to check if closure is static or not.
